### ISSUE TITLE
[BUG] Dependency injection not working with string dependencies

### ISSUE THREAD

**Issue by dingo-d** [infinum/eightshift-libs#154]
## Describe the bug
Looks like you cannot pass a string as a dependency.

## Steps to Reproduce
I have a tricky dependency problem. I'm building a custom submenu. For that purpose, I'm using a templated view class that will render my template (a PHP file to which I can pass attributes to).

The first problem was if I directly tried to add the class inside another class. I got a bunch of autowiring errors because php-di couldn't resolve the string (which is understandable). Then I added the dependency manually in my `Main.php`

```php
	protected function getServiceClasses(): array
	{
		return [
			BulkEmailOptions::class => [
				TemplatedView::class => [BulkEmailOptions::VIEW_URI]
			]
		];
	}
```

So my `BulkEmailOptions` class depends on a `TemplatedView` to which I need to pass a string defined in a constant `BulkEmailOptions::VIEW_URI`.

This falls in the `AbstractMain.php` on line 200

```php
	private function getDiDependencies(array $dependencies): array
	{
		return array_map(
			function ($dependency) {
				if (class_exists($dependency)) {
					return new Reference($dependency);
				}
				return $dependency;
			},
			$dependencies
		);
	}
```

Since my dependency is not a class, but a string the `class_exists` call will fall and I get a fatal error. I'd overwrite this implementation, but this part is `private` so no go.

> Uncaught Error: class_exists() expects parameter 1 to be string, array given
in /vendor/infinum/eightshift-libs/src/Main/AbstractMain.php on line 201

If I pass the dependency like this:

```php
	protected function getServiceClasses(): array
	{
		return [
			BulkEmailOptions::class => [
				TemplatedView::class => BulkEmailOptions::VIEW_URI
			]
		];
	}
```

I get 

> PHP Fatal error:  Uncaught Error: Cannot unpack array with string keys in /vendor/infinum/eightshift-libs/src/Main/AbstractMain.php:180
Stack trace:

We need to improve this part to allow cases where dependency is not a class, but any other primitive value (integer, string, boolean, float, array, or resource).

## Expected behavior
Dependency should be passed correctly in the manually defined dependency.

## Your Environment

* PHP version: 7.4.14
* Development environment: Valet
* Server type: Nginx
* Operating System and version: MacOS BigSur 11.2.2
* Libs version: 3.0.4

**Comment by iruzevic** [comment]
We have documented that this will not work

https://infinum.github.io/eightshift-docs/docs/basics/autowiring#what-if-my-class-has-a-primitive-parameter-string-int-etc-inside-a-constructor-method

**Comment by dingo-d** [comment]
Yes, then you need to manually pass the parameter in the `getServiceClasses` method. I think that the problem is that I have dependency inside a dependency inside a dependency.

```php
BulkEmailOptions::class => [
	TemplatedView::class => [BulkEmailOptions::VIEW_URI]
]
```

And we allow only 1 level dependency tree, e.g.

```php
BulkEmailOptions::class => [TemplatedView::class]
```

or

```php
BulkEmailOptions::class => ['Hi!']
```

**Comment by iruzevic** [comment]
We should write tests for the current implementation and then add additional fixes on the AW. agreed?

**Comment by dingo-d** [comment]
Yeah, I wouldn't go too deep, but this is really biting me in the ass atm, and I'm not sure how to resolve this implementation so that it works nicely in my project.

**Comment by IvanGrginovInf** [comment]
@dingo-d Yeah, we (I) should definitely write unit tests for this. Also dependencies are only resolved 2 lvls deep not 3. You don't need 3 levels (nor we ever used 3 lvls as far as I know)? Why? Because you can implement this by declaring the deepest dependency first. For example it should work like this (exactly like it did before when you manually had to write all dependencies):

```
TemplatedView::class => [
  BulkEmailOptions::VIEW_URI
],
BulkEmailOptions::class => [
  TemplatedView::class
],
```

**Comment by dingo-d** [comment]
The problem is: what if I need to pass something else to the templated view? Like I need another submenu page, which has templated view class as a dependency with a different dependency? In your case, the templated view will always use ` BulkEmailOptions::VIEW_URI`, and I need to change that based on the use case.

The main issue from what I can tell is that the autowiring and container seem to instantiate the classes and that's why the string typhinting fails (the DI container doesn't know what to put as a dependency just by looking at the class).

I tried to pass a DTO as a dependency for the `TemplatedView `, but in that case, the DTO failed in the DI 🤷🏼‍♂️ .

**Comment by IvanGrginovInf** [comment]
> I tried to pass a DTO as a dependency for the TemplatedView , but in that case, the DTO failed in the DI 🤷🏼‍♂️ .

Probably failed because autowiring was unable to build the DTO (maybe it had a primitive dependency). Tho autowiring should only wire services and dependencies of those services. And each of those dependencies needs to be built from something (i.e. you probably won't be able to autowire DTOs, which you construct from specific API responses or something)

**Comment by iruzevic** [comment]
@IvanGrginovInf is this fixed?

**Comment by IvanGrginovInf** [comment]
@iruzevic it should be fixed in 3.1.0, assuming you manually define the dependency tree for that class correctly in getServiceClasses (no nesting more than 1 level deep).

**Comment by iruzevic** [comment]
Ok I will close this issue then
